XForm/XForm/Commands/CosmosUnescape.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            if (context.Parser.HasAnotherPart)
+            {
+                throw new ArgumentOutOfRangeException($"{this.Verb} does not accept any arguments.");
+            }
+
+            return new CosmosUnescape(source);
+        }
+    }
+
+    public class CosmosUnescape : DataBatchEnumeratorWrapper
+    {
+        HashSet<int> _sourceColumns = new HashSet<int>();
+        Dictionary<String8, String8> _replacementMap;
+        String8 _bookmarkChar = String8.Convert("#", new byte[1]);
+
+        public CosmosUnescape(IDataBatchEnumerator source) : base(source)
+        {
+            for (int i = 0; i < source.Columns.Count; i++)
+            {
+                if (source.Columns[i].Type == typeof(String8))
+                {
+                    _sourceColumns.Add(i);
+                }
+            }
+        }
+
+        public Dictionary<String8, String8> ReplacementMap
+        {
+            get
+            {
+                if (_replacementMap == null)
+                {
+                    Dictionary<String8, String8> tempMap = new Dictionary<String8, String8>();
+                    tempMap.Add(String8.Convert("#R#", new byte[3]), String8.Convert("\r", new byte[1]));
+                    tempMap.Add(String8.Convert("#N#", new byte[3]), String8.Convert("\n", new byte[1]));
+                    tempMap.Add(String8.Convert("#TAB#", new byte[5]), String8.Convert("\t", new byte[1]));
+                    tempMap.Add(String8.Convert("#NULL#", new byte[6]), String8.Empty);
+                    tempMap.Add(String8.Convert("#HASH#", new byte[6]), String8.Convert("#", new byte[1]));
+
+                    _replacementMap = tempMap;
+                }
+
+                return _replacementMap;
+            }
+        }
+
+        public override Func<DataBatch> ColumnGetter(int columnIndex)
+        {
+            String8[] transformedArray = null;
+            String8Block block = new String8Block();
+
+            // Pass through columns other than the one being converted
+            if (!_sourceColumns.Contains(columnIndex)) return _source.ColumnGetter(columnIndex);
+
+            // Cache the function to get the source data
+            Func<DataBatch> sourceGetter = _source.ColumnGetter(columnIndex);
+
+            return () => Unescape(sourceGetter(), block, ref transformedArray);
+        }
+
+        private DataBatch Unescape(DataBatch batch, String8Block block, ref String8[] transformedArray)
+        {
+            block.Clear();
+            Allocator.AllocateToSize(ref transformedArray, batch.Count);
+
+            String8[] sourceArray = (String8[])batch.Array;
+            for (int i = 0; i < batch.Count; ++i)
+            {
+                String8 originalValue = sourceArray[batch.Index(i)];
+                transformedArray[i] = this.Unescape(originalValue, block);
+            }
+
+            return DataBatch.All(transformedArray, batch.Count);
```

Comment:
Hmm. We need to preserve nulls from the caller, too - check for a null array and set null if the source batch had a null. You'll want to write your own new null array because the source array may have indirection.